loki/transformations/inline.py

```diff
@@ -339,6 +360,89 @@ def inline_elemental_functions(routine):
     routine.spec = Transformer(import_map).visit(routine.spec)
 
 
+def recursive_inline_expression_map_update(expr_map, max_iterations=10):
+    """
+    Utility function to apply a inline substitution map for expressions to itself
+
+    Same logic used as in  :func:`recursive_expression_map_update`.
+                    
+    Parameters      
+    ----------      
+    expr_map : dict     
+        The inline substitution map that should be updated
+    max_iterations : int
+        Maximum number of iterations, corresponds to the maximum level of
+        nesting that can be replaced.
+    """
+    def apply_to_init_arg(name, arg, expr, mapper):
+        # Helper utility to apply the mapper only to expression arguments and
+        # retain the scope while rebuilding the node
+        if isinstance(arg, (tuple, Expression)):
+            return mapper(arg)
+        if name == 'scope':
+            return expr.scope
+        return arg
+
+    for _ in range(max_iterations):
+        # We update the expression map by applying it to the children of each replacement
+        # node, thus making sure node replacements are also applied to nested attributes,
+        # e.g. call arguments or array subscripts etc.
+        mapper = InlineSubstitutionMapper(expr_map) # SubstituteExpressionsMapper(expr_map)
+        prev_map, expr_map = expr_map, {
+            expr: type(replacement)(**{
+                name: apply_to_init_arg(name, arg, expr, mapper)
+                for name, arg in zip(replacement.init_arg_names, replacement.__getinitargs__())
+            })
+            for expr, replacement in expr_map.items()
+        }
+
+        # Check for early termination opportunities
+        if prev_map == expr_map:
+            break
+
+    return expr_map
+
+
+def inline_statement_functions(routine):
+    """
+    Replaces `InlineCall` expression to statement functions with the
+    called statement functions rhs.
+    """
+    # Keep track of removed symbols
+    removed_functions = set()
+
+    stmt_func_decls = FindNodes(StatementFunction).visit(routine.spec)
+    exprmap = {}
+    for call in FindInlineCalls().visit(routine.body):
+        if call.procedure_type is BasicType.DEFERRED:
+            continue
+        if call.procedure_type.is_function and isinstance(call.routine, StatementFunction):
```

Comment:
Just a little hint to avoid repeated lookup of type info:
```suggestion
        proc_type = call.procedure_type
        if proc_type is BasicType.DEFERRED:
            continue
        if proc_type.is_function and isinstance(proc_type.procedure, StatementFunction):
```